hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java

```diff
@@ -0,0 +1,159 @@
+			final Address mirrorReceiver,
+			final BigInteger userOfferedGasPrice,
+			final long maxGasAllowanceInTinybars,
+			final Account relayer,
+			final Id tokenId,
+			final WorldLedgers ledgers) {
+
+		HederaWorldState.Updater updater = (HederaWorldState.Updater) worldState.updater();
+		long intrinsicGas = gasCalculator.transactionIntrinsicGasCost(Bytes.EMPTY, false);
+		final MutableAccount mutableSender = TxProcessorUtil.getMutableSender(updater, sender);
+		final var isEthTx = relayer != null;
+		final long gasPrice = TxProcessorUtil.gasPriceTinyBarsGiven(
+				livePricesSource, consensusTime, isEthTx,
+				HederaFunctionality.ContractCall);
+		final Wei gasCost = Wei.of(Math.multiplyExact(gasLimit, gasPrice));
+		final Wei upfrontCost = gasCost.add(value);
+
+		var allowanceCharged = Wei.ZERO;
+		MutableAccount mutableRelayer = null;
+		if (isEthTx) {
+			mutableRelayer = TxProcessorUtil.getMutableRelayer(updater, relayer);
+		}
+		if (intrinsicGas > gasLimit) {
+			throw new InvalidTransactionException(INSUFFICIENT_GAS);
+		}
+		if (!isEthTx) {
+			TxProcessorUtil.chargeGas(gasCost, upfrontCost, mutableSender);
+		} else {
+			allowanceCharged = TxProcessorUtil.chargeForEth(
+					userOfferedGasPrice, gasCost, maxGasAllowanceInTinybars, mutableSender,
+					mutableRelayer, Wei.ZERO, gasPrice, gasLimit, value);
+		}
+
+		PrecompileMessage message = PrecompileMessage.builder()
+				.setLedgers(ledgers)
+				.setSenderAddress(sender.getId().asEvmAddress())
+				.setValue(Wei.of(value))
+				.setConsensusTime(consensusTime)
+				.setGasRemaining(gasLimit - intrinsicGas)
+				.setInputData(payload)
+				.setTokenID(tokenId.asGrpcToken())
+				.build();
+
+		callHtsPrecompile(message);
+		long gasUsedByTransaction = TxProcessorUtil.calculateGasUsedByTX(gasLimit, message,
+				dynamicProperties.maxGasRefundPercentage());
+
+		// return gas price to accounts
+		final long refunded = isEthTx ?
+				TxProcessorUtil.calculateRefundEth(gasLimit, gasPrice, mutableRelayer, allowanceCharged,
+						updater.getSbhRefund(), mutableSender, gasUsedByTransaction) :
+				TxProcessorUtil.calculateRefund(gasLimit, gasPrice,
+						gasUsedByTransaction, updater.getSbhRefund(), mutableSender);
+		// Send fees to coinbase
+		TxProcessorUtil.sendFeesToCoinbase(updater, dynamicProperties, gasLimit, gasPrice, refunded);
+
+		updater.commit();
+		final var isCompletedSuccess = message.getState() == COMPLETED_SUCCESS;
+		return TxProcessorUtil.buildTransactionResult(
+				isCompletedSuccess, message.getLogs(),
+				updater.getSbhRefund(), gasUsedByTransaction,
+				gasPrice, mirrorReceiver, message.getHtsOutputResult(),
+				message.getRevertReason(), message.getExceptionalHaltReason(),
+				Map.of());
+	}
+
+	private void callHtsPrecompile(PrecompileMessage message) {
+		htsPrecompiledContract.callHtsPrecompileDirectly(message);
+		final var gasRequirement = message.getGasRequired();
+
+		if (message.getGasRemaining() < gasRequirement) {
```

Comment:
Afaict this block is an exact duplicate of [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/contracts/execution/HederaMessageCallProcessor.java#L91); if we extract that to a static method and call here, we can fix missing code coverage on L#156.

Reply:
It would be great to remove the duplicated code here, unfortunately, MessageFrame and PrecomileMessage are not related by a common parent. Hence, such a method signature can't be applied. What comes to my mind as a possible solution is to introduce an additional model class to be passed as a first argument, instead of MessageFrame or an additional method with PrecompileMessage argument and a private method for the common logic shared by both methods. But I'm not sure if these approaches are appropriate here? I'm open to suggestions! 
:raised_hands: